### PR TITLE
[POC] Retrace backwards before compiling

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #92704

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

File: torch/_functorch/aot_autograd.py [hunk 1]
```diff
@@ -1869,12 +1869,28 @@ class CompiledFunctionBackward(torch.autograd.Function):
...
-                            )
+                        if config.use_dynamic_shapes:
+                            with context():
+                                # Retrace backwards with a new shape environment
+                                local_bw_module = make_fx(bw_module, tracing_mode="symbolic")(*all_args)
+                                with track_graph_compiling(aot_config, "backward"):
+                                    CompiledFunction.compiled_bw = aot_config.bw_compiler(
+                                        # NB: The None case is for when we
+                                        # have SymInt tangents, which are
+                                        # always passed with None
```
Comment (bdhirsh) [inline torch/_functorch/aot_autograd.py:1881]: dumb q: that comment seems to imply that the "SymInt tangent" case lines up with when `'val' not in n.meta`. But I thought that proxy nodes representing SymInts **do** get their `n.meta['val']` populated. Only "constant" nodes (plain numeric types/strings/output node/etc don't have that field)

**Comment by bdhirsh** [comment]
lmk know if I should wait before questions - what's the purpose of retracing the backward, instead of using the existing (partitioned) backward graph?

File: torch/_functorch/aot_autograd.py (hunk 1 above)
Comment (ezyang) [reply torch/_functorch/aot_autograd.py:1881]: It's not the SymInt input that is a problem (that indeed has a meta), but the *tangent* of that input, which is None and doesn't have a meta.

TBH though these Nones are tiresome and I kind of would like it if we could change the calling convention so SymInt inputs don't get tangents

**Comment by ezyang** [comment]
There are a few reasons.

The proximal reasoning goes like this:

* I would like to pass fake tensors when compiling backwards
* When I pass the fake tensors that I used when I originally traced forwards-backwards, I have an Inductor compilation problem: sometimes I refer to shape variables but they are never passed in as an input! For example, suppose I save a tensor with size s0 * 2 for backwards. If s0 never shows up by itself in any of the other inputs for backwards, I actually have no way of directly instantiating s0 for backwards
* If I retrace, I allocate fresh size variables for all the inputs, so s0 * 2 gets reassigned to s1 and everything is great.

Note that doing it this way loses guard information, so we need to propagate guards to backwards shape env (not implemented yet.)

There is also a long term reason to retrace. Our backward graphs are technically unsound as we trace them assuming grad is contiguous, but it actually might not be contiguous. We do not know what the grad stride is until we actually hit backwards; only then can we trace! So morally we need a two level cache and we need to trace at backwards time.

**Pull request closed without merging**